File: frontend/src/api/sms.ts
import api from './client'

export interface SmsMessage {
  id: number
  modem_id: number
  direction: 'inbound' | 'outbound'
  phone_number: string
  content: string
  status: 'pending' | 'sent' | 'failed' | 'received'
  error_message: string | null
  sent_at: string | null
  received_at: string | null
  created_at: string
}

export interface SmsTemplate {
  id: number
  name: string
  content: string
  variables: string[] | null
  created_at: string
}

export interface ScheduledTask {
  id: number
  name: string
  modem_id: number
  recipients: string[]
  content: string
  cron_expression: string | null
  send_once_at: string | null
  status: 'active' | 'paused' | 'completed' | 'failed'
  last_run_at: string | null
  next_run_at: string | null
  run_count: number
  created_at: string
}

export const sendSmsApi = (data: { modem_id: number; phone_number: string; content: string }) =>
  api.post<SmsMessage>('/sms/send', data)

export const getMessagesApi = (params?: { modem_id?: number; direction?: string; skip?: number; limit?: number }) =>
  api.get<SmsMessage[]>('/sms/messages', { params })

export const getTemplatesApi = () => api.get<SmsTemplate[]>('/sms/templates')
export const createTemplateApi = (data: { name: string; content: string; variables?: string[] }) =>
  api.post<SmsTemplate>('/sms/templates', data)
export const deleteTemplateApi = (id: number) => api.delete(`/sms/templates/${id}`)

export const getTasksApi = () => api.get<ScheduledTask[]>('/sms/tasks')
export const createTaskApi = (data: Partial<ScheduledTask>) => api.post<ScheduledTask>('/sms/tasks', data)
export const updateTaskApi = (id: number, data: Partial<ScheduledTask>) => api.patch<ScheduledTask>(`/sms/tasks/${id}`, data)
export const deleteTaskApi = (id: number) => api.delete(`/sms/tasks/${id}`)
export const runTaskNowApi = (id: number) => api.post(`/sms/tasks/${id}/run-now`)

File: backend/app/schemas/sms.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Any
from app.models.sms import SmsDirection, SmsStatus, TaskStatus


class SmsSendRequest(BaseModel):
    modem_id: int
    phone_number: str
    content: str


class SmsMessageOut(BaseModel):
    id: int
    modem_id: int
    direction: SmsDirection
    phone_number: str
    content: str
    status: SmsStatus
    error_message: Optional[str]
    sent_at: Optional[datetime]
    received_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


class SmsTemplateCreate(BaseModel):
    name: str
    content: str
    variables: Optional[List[str]] = None


class SmsTemplateOut(BaseModel):
    id: int
    name: str
    content: str
    variables: Optional[List[str]]
    created_at: datetime

    model_config = {"from_attributes": True}


class ScheduledTaskCreate(BaseModel):
    name: str
    modem_id: int
    recipients: List[str]
    content: str
    cron_expression: Optional[str] = None
    send_once_at: Optional[datetime] = None


class ScheduledTaskUpdate(BaseModel):
    name: Optional[str] = None
    recipients: Optional[List[str]] = None
    content: Optional[str] = None
    cron_expression: Optional[str] = None
    send_once_at: Optional[datetime] = None
    status: Optional[TaskStatus] = None


class ScheduledTaskOut(BaseModel):
    id: int
    name: str
    modem_id: int
    recipients: List[Any]
    content: str
    cron_expression: Optional[str]
    send_once_at: Optional[datetime]
    status: TaskStatus
    last_run_at: Optional[datetime]
    next_run_at: Optional[datetime]
    run_count: int
    created_at: datetime
    created_by_id: Optional[int] = None
    created_by_username: Optional[str] = None

    model_config = {"from_attributes": True}


class TaskStatsOut(BaseModel):
    total: int
    active: int
    paused: int
    completed: int
    failed: int
